core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1207 @@
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.42000000000000004,
+        xLabel: 0.1625,
+        yLabel: 0.23333333333333334,
+        id: 'State2',
+        label: 'State2',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State3: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.42000000000000004,
+        xLabel: 0.3875,
+        yLabel: 0.5,
+        id: 'State3',
+        label: 'State3',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State4: {
+        depth: 2,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.6866666666666666,
+        xLabel: 0.1625,
+        yLabel: 0.7666666666666666,
+        id: 'State4',
+        label: 'State4',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+
+    expect(sgls.getAugmentedLinks(nodeData, links2)).toBeUndefined();
+  });
+
+  it('should get correct graph width and height', () => {
+    let nodeData = {
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.42000000000000004,
+        xLabel: 0.1625,
+        yLabel: 0.5,
+        id: 'State1',
+        label: 'State1',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.15333333333333335,
+        xLabel: 0.1625,
+        yLabel: 0.23333333333333334,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 2,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.6866666666666666,
+        xLabel: 0.1625,
+        yLabel: 0.7666666666666666,
+        id: 'End',
+        label: 'End',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.42000000000000004,
+        xLabel: 0.3875,
+        yLabel: 0.5,
+        id: 'State2',
+        label: 'State2',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+
+    let graphWidthUpperBound = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+
+    expect(graphWidthUpperBound).toBe(630);
```

Comment:
Can you add some explanation (in comments) of where these numbers come from and why they are correct? Test files should be understandable in a self-contained way.

Reply:
Done

Reply:
I see no comments here?

Reply:
Sorry, I missed this one as it was same comment as for the next test. Added

Reply:
Updated